src/oop/special_class_method/hard/q3/MainTest.java

```diff
@@ -0,0 +1,245 @@
+                {updatedValue, String.class}
+        };
+        String[] setMethodModifiers = {"public"};
+        String incorrectSetMethodMessage = String.join(" ",
+                "Your", setStringMethodName, "method does not correctly set the value of the", stringFieldName, "attribute.");
+        Object setMethodOutput = testClass.callMethod(setStringMethodName, setMethodArguments, setMethodModifiers, classInstance);
+        _assertEquals(updatedValue, testClass.getFieldValue(classInstance, stringFieldName), incorrectSetMethodMessage);
+        assertNull(setMethodOutput, String.join(" ", "Your", setStringMethodName, "should not return any output"));
+    }
+
+    private static Stream<String> stringInputProvider() {
+        return Stream.of("Tall", "Red", "Horizontal");
+    }
+
+    @ParameterizedTest
+    @MethodSource("stringInputProvider")
+    public void chairClassHasCorrectGetSoftnessMethod(String value) throws Throwable {
+        Object[][] arguments = {
+                {Math.PI, double.class},
+                {value, String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        String[] getMethodModifiers = {"public"};
+        String incorrectGetMethodMessage = String.join(" ",
+                "Your", getStringMethodName, "method does not correctly get the value of the", stringFieldName, "attribute.");
+        Object getMethodOutput = testClass.callMethod(getStringMethodName, getMethodModifiers, classInstance);
+        _assertEquals(value, getMethodOutput, incorrectGetMethodMessage);
+    }
+
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void chairClassHasCorrectToStringMethod(double value, String b) throws Throwable {
+        Object[][] arguments = {
+                {value, double.class},
+                {b, String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        String[] methodModifiers = {"public"};
+        String expected = "This " + className + "'s " + doubleFieldName + " is " + value + ", and it's " + stringFieldName + " is " + b;
+        String incorrectToStringMessage = String.join(" ",
+                "Your", className, " toString method does not return the correct String.");
+        Object output = testClass.callMethod("toString", methodModifiers, classInstance);
+        _assertEquals(expected, output, incorrectToStringMessage);
+    }
+
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void chairClassMethodsWorkTogether(double value, String b) throws Throwable {
+        double initialValue = Math.random() * 100;
+        Object[][] arguments = {
+                {initialValue, double.class},
+                {"First String", String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        _assertEquals(initialValue, testClass.callMethod(getDoubleMethodName, classInstance),
```

Comment:
no need to test getting before set is called. Aside from that it's good for approval